libsolidity/formal/BMC.cpp

```diff
@@ -0,0 +1,885 @@
+	if (!_node.isDoWhile())
+		_node.condition().accept(*this);
+
+	mergeVariables(touchedVars, expr(_node.condition()), indicesAfterLoop, copyVariableIndices());
+
+	m_loopExecutionHappened = true;
+	return false;
+}
+
+// Here we consider the execution of two branches similar to WhileStatement.
+bool BMC::visit(ForStatement const& _node)
+{
+	if (_node.initializationExpression())
+		_node.initializationExpression()->accept(*this);
+
+	auto indicesBeforeLoop = copyVariableIndices();
+
+	// Do not reset the init expression part.
+	auto touchedVars = touchedVariables(_node.body());
+	if (_node.condition())
+		touchedVars += touchedVariables(*_node.condition());
+	if (_node.loopExpression())
+		touchedVars += touchedVariables(*_node.loopExpression());
+
+	m_context.resetVariables(touchedVars);
+
+	if (_node.condition())
+	{
+		_node.condition()->accept(*this);
+		if (isRootFunction())
+			addVerificationTarget(
+				VerificationTarget::Type::ConstantCondition,
+				expr(*_node.condition()),
+				_node.condition()
+			);
+	}
+
+	m_context.pushSolver();
+	if (_node.condition())
+		m_context.addAssertion(expr(*_node.condition()));
+	_node.body().accept(*this);
+	if (_node.loopExpression())
+		_node.loopExpression()->accept(*this);
+	m_context.popSolver();
+
+	auto indicesAfterLoop = copyVariableIndices();
+	// We reset the execution to before the loop
+	// and visit the condition.
+	resetVariableIndices(indicesBeforeLoop);
+	if (_node.condition())
+		_node.condition()->accept(*this);
+
+	auto forCondition = _node.condition() ? expr(*_node.condition()) : smt::Expression(true);
+	mergeVariables(touchedVars, forCondition, indicesAfterLoop, copyVariableIndices());
+
+	m_loopExecutionHappened = true;
+	return false;
+}
+
+void BMC::endVisit(UnaryOperation const& _op)
+{
+	SMTEncoder::endVisit(_op);
+
+	if (_op.annotation().type->category() == Type::Category::RationalNumber)
+		return;
+
+	switch (_op.getOperator())
+	{
+	case Token::Inc: // ++ (pre- or postfix)
+	case Token::Dec: // -- (pre- or postfix)
+		addVerificationTarget(
+			VerificationTarget::Type::UnderOverflow,
+			expr(_op),
+			&_op
+		);
+		break;
+	case Token::Sub: // -
+		if (_op.annotation().type->category() == Type::Category::Integer)
+			addVerificationTarget(
+				VerificationTarget::Type::UnderOverflow,
+				expr(_op),
+				&_op
+			);
+		break;
+	default:
+		break;
+	}
+}
+
+void BMC::endVisit(FunctionCall const& _funCall)
+{
+	solAssert(_funCall.annotation().kind != FunctionCallKind::Unset, "");
+	if (_funCall.annotation().kind != FunctionCallKind::FunctionCall)
+	{
+		SMTEncoder::endVisit(_funCall);
+		return;
+	}
+
+	FunctionType const& funType = dynamic_cast<FunctionType const&>(*_funCall.expression().annotation().type);
+	switch (funType.kind())
+	{
+	case FunctionType::Kind::Assert:
+		SMTEncoder::endVisit(_funCall);
+		visitAssert(_funCall);
+		break;
+	case FunctionType::Kind::Require:
+		visitRequire(_funCall);
+		SMTEncoder::endVisit(_funCall);
```

Comment:
Here we need to verify the constant condition before adding the require condition as a constraint in `SMTEncoder::endVisit(_funCall)`.